language/diem-tools/prefetched-transaction-replay/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+
+        Ok(PrefetchTransactionReplayer {
+            state,
+            debugger,
+            rpc_client,
+        })
+    }
+
+    fn get_keys_read_for_tx(
+        &self,
+        tx: &Transaction,
+        modules: &[CompiledModule],
+    ) -> Result<Vec<ResourceKey>> {
+        let signed_tx = match tx {
+            Transaction::UserTransaction(u) => u,
+            _ => bail!("can only process user transactions"),
+        };
+
+        let rw_analysis = ReadWriteSetAnalysis::new(read_write_set::analyze(modules)?);
+        rw_analysis.get_keys_read(signed_tx, &self.state.view)
+    }
+
+    fn get_unique_addresses_from_keys(&self, key_reads: &[ResourceKey]) -> Vec<AccountAddress> {
+        key_reads
+            .iter()
+            .map(|k| k.address)
+            .collect::<HashSet<_>>()
+            .into_iter()
+            .collect()
+    }
+
+    fn extract_account_state(
+        &self,
+        state_with_proof: &AccountStateWithProofView,
+    ) -> Option<AccountState> {
+        // NOTE: if this actually has a chance to fail, it might be better to change
+        // the function to return Result<Option<AccountState>, _>
+        state_with_proof.blob.as_ref().and_then(|bytes| {
+            bcs::from_bytes::<AccountStateBlob>(bytes)
+                .map_err(|e| anyhow!(e.to_string()))
+                .and_then(|account_state_blob| AccountState::try_from(&account_state_blob))
+                .ok()
+        })
+    }
+
+    fn fetch_accounts_storage(
+        &self,
+        addresses: &[AccountAddress],
+    ) -> Result<Vec<AccountAddressWithState>> {
+        let requests = addresses
+            .iter()
+            .map(|a| MethodRequest::get_account_state_with_proof(*a, None, None))
+            .collect();
+
+        let results = self
+            .rpc_client
+            .batch(requests)?
+            .into_iter()
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let account_states: Vec<AccountAddressWithState> = results
+            .into_iter()
+            .zip(addresses.iter())
+            .filter_map(|(r, a)| match r.into_inner() {
+                MethodResponse::GetAccountStateWithProof(v) => self
+                    .extract_account_state(&v)
+                    .map(|state| AccountAddressWithState { address: *a, state }),
+                _ => None,
+            })
+            .collect();
+
+        Ok(account_states)
+    }
+
+    fn populate_storage_for_transactions(
+        &self,
+        txs: &[Transaction],
+        version: Option<u64>,
+    ) -> Result<()> {
+        let replay_version = match version {
+            Some(v) => v,
+            None => self.debugger.get_latest_version()?,
+        };
+
+        let modules = self
+            .debugger
+            .get_diem_framework_modules_at_version(replay_version, true)?;
+
+        let mut keys = vec![];
+        for tx in txs.iter() {
+            let new_keys = self.get_keys_read_for_tx(tx, &modules)?;
```

Comment:
This will re-run the read/write summaries for every tx (ok) *and* its dependencies (ok, but potentially slow). It would be great if when analyzing tx2, we could leverage the summaries computed for dependencies of tx1. I'm not sure if the API's exposed for the read/write set analysis make this easy, but we should be able to tweak them.

One way we could do this is by analyzing `modules` first and then passing the result along to each tx (perhaps even in parallel if it would help).

Reply:
Right, that makes sense! I'll look into this!